documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+
+### Inner parts of a test
+
+While writing test cases, it will be easier to think of the test case's contents as if there were three clear divisions:
+
+- The first lines of the test should deal with setting up the necessary variables and values that will be used for the main functionality that will be tested.
+- The body of the test should focus on using these values to trigger the main functionality that is being tested.
+- The last set of lines should be focused on comparing the results obtained, to confirm that they are in the expected shape, or contain the expected values.
+
+### Using conditionals
+
+Ideally, each test case should only have one possible behavior. To minimize unexpected outcomes, conditionals should be avoided. The only case in which conditionals are encouraged is to specify when to skip a test. Some tests might only run in a browser, or only run in NodeJS, or only run live (no record or playback), for this purpose, conditionals can be used. For example:
+
+```ts
+import { isNode } from "@azure/core-http";
+
+describe("testing the client's basic methods", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A #browser", function() {
+    if (isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test B #node", function() {
+    if (!isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test C #live", function() {
+    if (process.env.TEST_MODE) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+});
+```
+
+In case conditionals might appear to be necessary in other scenarios, consider separating the test case into as many test cases as necessary first.
+
+### Using delays
+
+The API methods that we provide should always provide a way for users to know when an operation finishes. For this reason, any method that doesn't immediately respond with a completed operation should use our [Long Running Operations strategy (core-lro)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro).
+
+If any method can't use `core-lro` within reasonable time, and thus we are inclined to use a different delay strategy, we should always wait until the next possible operation is can be fulfilled.
+
+For example, in KeyVault-Keys we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
+
+```ts
+import { delay } from "@azure/test-utils-recorder";
+// ...
+
+describe("Keys client - restore keys and recover backups", () => {
+  // ...
+  it("can restore a key with a given backup", async function() {
+    // ...
+    await client.createKey(keyName, "RSA");
+    const backup = await client.backupKey(keyName);
+    await client.purgeDeletedKey(keyName);
+    while (true) {
```

Comment:
So in EH, where tests could potentially take an indeterminate amount of time (by accident) we have a loopUntil() method that basically retries a set # of times with a delay so that way you don't run eternally.

Would you suggest something along that nature here? (at the very least, a counter?)

Reply:
That's a great point and that's why I submitted the while, to get smarter responses. We did have a retry tool, but these tools need to be moved to a common test utilities package, so that other packages can use it.

We do use whiles for now since our retry utility was bugged and would get the test stuck forever. In theory, mocha's timeout should handle the exceptional case in which we might not get resolved on an appropriate time, though I do agree that a common utility will help. We should think together on a common set of test utilities. That work is perhaps separate from the initial test guidelines approach (or not! Up for debate).

Reply:
After some thought, we should not have things like my retry tool, or loopUntil, at all. These tests can exist as temporary measures and not run on live test runs. All of our methods that don't return instantly should use `core-lro`. If we can't make them use core-lro because of a service limitation, we should report this issue to the service team and continue pushing until they provide us an approach that we can reasonably make into a LRO.